File: problems/quick-sort/quick-sort.py
from random import randint

def quickSort(lst):
    # List of 0 or 1 items is already sorted
    if len(lst) <= 1:
        return lst
    else:
        # Pivot can be chosen randomly
        pivotIndex = randint(0, len(lst)-1)
        pivot = lst[pivotIndex]
        # Elements lower than and greater than pivot
        lesser, greater = [], []

        for index in range(len(lst)):
            # Don't do anything if you're at the pivot
            if index == pivotIndex:
                pass
            else:
                # Sort elements into < pivot and >= pivot
                el = lst[index]
                if el < pivot:
                    lesser.append(el)
                else:
                    greater.append(el)
                    
        # Sort lesser and greater, concatenate results
        return quickSort(lesser) + [pivot] + quickSort(greater)

File: problems/quick-sort/quick-sort.js
module.exports = function quickSort (array, compare) {
  var lesser  = [],
      greater = [],
      pivot;

  // Not an array, empty or array of 1 is already sorted
  if (!Array.isArray(array) || array.length < 2) {
    return array;
  }

  // Create a compare func if not passed in
  if (typeof compare !== 'function') {
    compare = function (a, b) {
      return a > b ? 1 : -1;
    };
  }

  // Get our pivot, this can be random
  pivot = array.splice(~~(Math.random() * array.length), 1);

  // Iterate and put vals into either lesser or greater lists compared
  // to the pivot
  for (var i = 0; i < array.length; i++) {
    if (compare(array[i], pivot) < 1) {
      lesser.push(array[i]);
    } else {
      greater.push(array[i]);
    }
  }

  // Sort lesser and greater lists, concat results
  return Array.prototype.concat(
    quickSort(lesser, compare),
    pivot,
    quickSort(greater, compare)
  );
};
